src/server.rs

```diff
@@ -0,0 +1,352 @@
+                                }),
+            }
+        }
+    }
+}
+
+pub struct Inner {
+    env: Arc<Environment>,
+    server: *mut GrpcServer,
+    bind_addrs: Vec<(String, u32)>,
+    slots_per_cq: usize,
+    shutdown: AtomicBool,
+    handlers: HashMap<&'static [u8], Handler>,
+}
+
+impl Inner {
+    /// Get the handler for the requested method path.
+    pub fn get_handler(&self, method: &[u8]) -> Option<&Handler> {
+        self.handlers.get(method)
+    }
+}
+
+/// Request notification of a new call.
+pub fn request_call(inner: Arc<Inner>, cq: &CompletionQueue) {
+    if inner.shutdown.load(Ordering::Relaxed) {
+        return;
+    }
+    let server_ptr = inner.server;
+    let prom = CallTag::request(inner);
+    let request_ptr = prom.request_ctx().unwrap().as_ptr();
+    let prom_box = Box::new(prom);
+    let tag = Box::into_raw(prom_box);
+    let code = unsafe {
+        grpc_sys::grpcwrap_server_request_call(server_ptr, cq.as_ptr(), request_ptr, tag as *mut _)
+    };
+    if code != GrpcCallStatus::Ok {
+        Box::from(tag);
+        panic!("failed to request call: {:?}", code);
+    }
+}
+
+/// An asynchronize shutdown future.
+pub struct ShutdownFuture {
+    cq_f: CqFuture<()>,
+}
+
+impl Future for ShutdownFuture {
+    type Item = ();
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<(), Error> {
+        try_ready!(self.cq_f.poll());
+        Ok(Async::Ready(()))
+    }
+}
+
+// It's safe to request call simultaneously.
+unsafe impl Sync for Inner {}
+unsafe impl Send for Inner {}
+
+pub struct Server {
+    inner: Arc<Inner>,
+}
+
+impl Server {
+    /// Shutdown the server asynchronously.
+    pub fn shutdown(&mut self) -> ShutdownFuture {
+        let (cq_f, prom) = CallTag::shutdown_pair();
+        let prom_box = Box::new(prom);
+        let tag = Box::into_raw(prom_box);
+        unsafe {
+            let cq_ptr = self.inner.env.completion_queues()[0].as_ptr();
+            grpc_sys::grpc_server_shutdown_and_notify(self.inner.server, cq_ptr, tag as *mut _)
+        }
+        self.inner.shutdown.store(true, Ordering::SeqCst);
+        ShutdownFuture { cq_f: cq_f }
+    }
+
+    /// Cancel all in-progress calls.
+    ///
+    /// Only usable after shutdown.
+    pub fn cancel_all_calls(&mut self) {
+        unsafe { grpc_sys::grpc_server_cancel_all_calls(self.inner.server) }
+    }
+
+    /// Start a server.
+    ///
+    /// Tells all listeners to start listening.
+    pub fn start(&mut self) {
+        unsafe {
+            grpc_sys::grpc_server_start(self.inner.server);
+            for cq in self.inner.env.completion_queues() {
+                for _ in 0..self.inner.slots_per_cq {
+                    request_call(self.inner.clone(), cq);
+                }
+            }
+        }
+    }
+
+    /// Get the binded addresses.
+    pub fn bind_addrs(&self) -> &[(String, u32)] {
+        &self.inner.bind_addrs
+    }
+}
+
+impl Drop for Server {
```

Comment:
It won't. Because once the shutdown future is resolved, there will be no further request_call hence it's safe to drop the server. But I will move it to Inner for best practice.